microcosm_pubsub/producer.py

```diff
@@ -85,40 +78,37 @@ def create_message(self, media_type, dct, uri=None, opaque_data=None, **kwargs)
 
         opaque_data[PUBLISHED_KEY] = str(time())
 
-        topic_arn = self.choose_topic_arn(media_type)
-
-        message_attributes = self.choose_message_attributes(media_type)
-
         message = self.pubsub_message_schema_registry.find(media_type).encode(
             dct,
             opaque_data=opaque_data,
             media_type=media_type,
             uri=uri,
             **kwargs
         )
-        return PubsubMessage(
+        return PublishedMessage(
             media_type=media_type,
             message=message,
-            message_attributes=message_attributes,
             opaque_data=opaque_data,
-            topic_arn=topic_arn,
         )
 
-    def publish_message(self, pubsub_message: PubsubMessage):
+    def publish_message(self, message: PublishedMessage):
         extra = dict(
-            media_type=pubsub_message.media_type,
-            **pubsub_message.opaque_data
+            media_type=message.media_type,
+            **message.opaque_data
         )
         self.logger.debug("Publishing message with media type {media_type}", extra=extra)
         publish_result = "SUCCESS"
         publish_exception = None
 
+        # Run this before `try` block to avoid TopicNotDefinedError being swallowed
+        topic_arn = self.choose_topic_arn(message.media_type)
+
         with elapsed_time(extra):
             try:
                 result = self.sns_client.publish(
-                    TopicArn=pubsub_message.topic_arn,
-                    Message=pubsub_message.message,
-                    MessageAttributes=pubsub_message.message_attributes,
+                    TopicArn=topic_arn,
+                    Message=message.message,
+                    MessageAttributes=self.choose_message_attributes(message.media_type),
```

Comment:
We can compute `TopicArn` and `MessageAttributes` here, instead of putting them on the pubsub message. Both are SNS-specific.